voc/python/ast.py

```diff
@@ -986,27 +995,71 @@ def _create_function(self, node, func_name, decorator_list):
 
         # Now actually define the function.
         for decorator in decorator_list:
-            self.visit(decorator)
-            self.context.add_opcodes(
-                JavaOpcodes.CHECKCAST('org/python/Callable'),
-                java.Array(1),
-                JavaOpcodes.DUP(),
-                JavaOpcodes.ICONST_0(),
-            )
+            # The @super decorator on __init__() is a special case. It tells us how
```

Comment:
On closer inspection, I'm not sure it does help. The call to push/pop context happens quite deep in this block, and has to happen *after* the `InitMethod` is constructed. This means you have to pass the visitor to the klass instance, which would be first place in the code this happens, and would also introduce a circular dependency. 

I've added a minor change to improve error reporting,  but unless you can see a better way to factor this out, I'm not sure it's possible to move the logic to `klass.py`